### PR TITLE
[#901] fix(server): respect disk real usage when checking whether local storage reaches watermark limit

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

Respect disk real usage when checking whether local storage reaches watermark limit

### Why are the changes needed?

Fix: #901

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

1. UTs

### CODE REVIEW

**Comment by zuston** [comment]
Could you help review this? @awdavidson

**Comment by zuston** [comment]
PTAL @jerqi @xianjingfeng

File: server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java [hunk 1]
```diff
@@ -211,6 +230,27 @@ boolean checkIsSpaceEnough() {
       return isHealthy;
     }
 
+    boolean checkWatermarkLimit() {
+      double usagePercent = getWholeDiskUsedSpace(storageDir) * 100.0 / getTotalSpace(storageDir);
```
Comment (xianjingfeng) [inline server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java:234]: `getTotalSpace` has been invoked in `getWholeDiskUsedSpace`, so could we avoid invoking it twice?

Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java:234]: Nice catch! I think it can

**Comment by xianjingfeng** [comment]
Should we remove `rss.server.disk.capacity`?

**Comment by zuston** [comment]
> Should we remove `rss.server.disk.capacity`?

When using the whole disk, capacity is not specified explictly.

But when colocation or we hope uniffle uses the limited disk, the capacity is also needed. Right? 

If so, the watermark limit check should be based on two ways, including wholeUsedDisk/wholeDiskSize and uniffleUsedDisk/uniffleDiskCapacity ? Right @xianjingfeng

**Comment by xianjingfeng** [comment]
> > Should we remove `rss.server.disk.capacity`?
> 
> When using the whole disk, capacity is not specified explictly.
> 
> But when colocation or we hope uniffle uses the limited disk, the capacity is also needed. Right?
> 
> If so, the watermark limit check should be based on two ways, including wholeUsedDisk/wholeDiskSize and uniffleUsedDisk/uniffleDiskCapacity ? Right @xianjingfeng

Right

**Comment by zuston** [comment]
But I have no idea how to take the compatible way to implement?

**Comment by xianjingfeng** [comment]
> But I have no idea how to take the compatible way to implement?

uniffleUsedDisk=du ${uniffle_data_dir} ?
And add an option for choosing which way to check this disk?

**Comment by zuston** [comment]
> > But I have no idea how to take the compatible way to implement?
> 
> uniffleUsedDisk=du ${uniffle_data_dir} ? And add an option for choosing which way to check the disk?

I think the unified way will be better. Like this? 

For high watermark check logic: 

`if (uniffleUsedDiskSize > uniffleSpecifiedCapacity * highWatermarkRatio || WholeDiskUsed > wholeDiskSize * highWatermarkRatio)` 

For the low watermark check logic:

`if (uniffleUsedDiskSize < uniffleSpecifiedCapacity * lowWatermarkRatio && WholeDiskUsed < wholeDiskSize * lowWatermarkRatio)` 


WDYT? @xianjingfeng 
CC @jerqi

**Comment by jerqi** [comment]
> > > But I have no idea how to take the compatible way to implement?
> > 
> > 
> > uniffleUsedDisk=du ${uniffle_data_dir} ? And add an option for choosing which way to check the disk?
> 
> I think the unified way will be better. Like this?
> 
> For high watermark check logic:
> 
> `if (uniffleUsedDiskSize > uniffleSpecifiedCapacity * highWatermarkRatio || WholeDiskUsed > wholeDiskSize * highWatermarkRatio)`
> 
> For the low watermark check logic:
> 
> `if (uniffleUsedDiskSize < uniffleSpecifiedCapacity * lowWatermarkRatio && WholeDiskUsed < wholeDiskSize * lowWatermarkRatio)`
> 
> WDYT? @xianjingfeng CC @jerqi

Do we have an intreface  DiskManagePolicy? Users can use different policies by different configuration options.

**Comment by awdavidson** [comment]
Apologies a little late to this. I agree with keeping ‘rss.server.disk.capacity’ for now. This setting is useful for making uniffle aware of the disk allocated when deploying uniffle on Kubernetes using a volume with a size limit (mainly for colocation as mentioned).

As for the watermark check, I would say yes, when a disk capacity is configured we should check against the disk size and uniffle disk used capacity

**Comment by zuston** [comment]
> As for the watermark check, I would say yes, when a disk capacity is configured we should check against the disk size and uniffle disk used capacity

+1. I will push this PR forward by this way.

**Comment by zuston** [comment]
new PR to implement this. https://github.com/apache/incubator-uniffle/pull/1337 @awdavidson @jerqi @xianjingfeng PTAL

**Pull request closed without merging**